File: hallucinate_app/python/hallucinate_app/test/test_github_issue_reporter.py
"""
Test GitHub Issue Reporter

Tests the functionality of the GitHub issue reporter.
"""

import os
import sys
import asyncio
import unittest
from datetime import datetime

# Add parent directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))

from hallucinate_app.github_issue_reporter import (
    GitHubIssueReporter,
    IssueReportConfig,
    get_reporter,
    report_error
)
from hallucinate_app.error_monitor import ErrorData, ErrorLevel, ErrorSource


class TestGitHubIssueReporter(unittest.TestCase):
    """Test GitHub issue reporter functionality"""
    
    def setUp(self):
        """Set up test fixtures"""
        # Create a test configuration with dry_run enabled
        self.config = IssueReportConfig(
            enabled=True,
            github_token='test_token',
            repository='test_owner/test_repo',
            min_error_level=ErrorLevel.ERROR,
            dry_run=True  # Don't actually create issues during tests
        )
        
        self.reporter = GitHubIssueReporter(self.config)
    
    def test_config_from_env(self):
        """Test configuration from environment variables"""
        # Set environment variables
        os.environ['GITHUB_ISSUE_REPORTER_ENABLED'] = 'true'
        os.environ['GITHUB_TOKEN'] = 'test_token'
        os.environ['GITHUB_REPOSITORY'] = 'owner/repo'
        os.environ['GITHUB_ISSUE_MIN_LEVEL'] = 'warning'
        os.environ['GITHUB_ISSUE_RATE_LIMIT'] = '5'
        os.environ['GITHUB_ISSUE_DRY_RUN'] = 'true'
        
        config = IssueReportConfig.from_env()
        
        self.assertTrue(config.enabled)
        self.assertEqual(config.github_token, 'test_token')
        self.assertEqual(config.repository, 'owner/repo')
        self.assertEqual(config.min_error_level, ErrorLevel.WARNING)
        self.assertEqual(config.rate_limit_per_hour, 5)
        self.assertTrue(config.dry_run)
        
        # Clean up
        for key in ['GITHUB_ISSUE_REPORTER_ENABLED', 'GITHUB_TOKEN', 'GITHUB_REPOSITORY',
                    'GITHUB_ISSUE_MIN_LEVEL', 'GITHUB_ISSUE_RATE_LIMIT', 'GITHUB_ISSUE_DRY_RUN']:
            os.environ.pop(key, None)
    
    def test_error_fingerprint(self):
        """Test error fingerprinting"""
        error1 = ErrorData(
            id='test-1',
            timestamp=datetime.now().isoformat(),
            level=ErrorLevel.ERROR,
            source=ErrorSource.PYTHON,
            component='TestComponent',
            operation='test_operation',
            message='Test error message'
        )
        
        error2 = ErrorData(
            id='test-2',
            timestamp=datetime.now().isoformat(),
            level=ErrorLevel.ERROR,
            source=ErrorSource.PYTHON,
            component='TestComponent',
            operation='test_operation',
            message='Test error message'
        )
        
        error3 = ErrorData(
            id='test-3',
            timestamp=datetime.now().isoformat(),
            level=ErrorLevel.WARNING,
            source=ErrorSource.PYTHON,
            component='TestComponent',
            operation='test_operation',
            message='Different error message'
        )
        
        fingerprint1 = self.reporter.generate_error_fingerprint(error1)
        fingerprint2 = self.reporter.generate_error_fingerprint(error2)
        fingerprint3 = self.reporter.generate_error_fingerprint(error3)
        
        # Same errors should have same fingerprint
        self.assertEqual(fingerprint1, fingerprint2)
        
        # Different errors should have different fingerprints
        self.assertNotEqual(fingerprint1, fingerprint3)
    
    def test_duplicate_detection(self):
        """Test duplicate error detection"""
        fingerprint = 'test_fingerprint'
        
        # Initially not a duplicate
        self.assertFalse(self.reporter.is_duplicate(fingerprint))
        
        # Mark as reported
        self.reporter.reported_fingerprints.add(fingerprint)
        self.reporter.fingerprint_timestamps[fingerprint] = datetime.now().timestamp()
        
        # Should now be detected as duplicate
        self.assertTrue(self.reporter.is_duplicate(fingerprint))
    
    def test_rate_limiting(self):
        """Test rate limiting"""
        # Initially should pass rate limit
        self.assertTrue(self.reporter.check_rate_limit())
        
        # Add many recent issues
        import time
        now = time.time()
        for i in range(self.config.rate_limit_per_hour):
            self.reporter.issue_count_per_hour.append(now)
        
        # Should now fail rate limit
        self.assertFalse(self.reporter.check_rate_limit())
    
    def test_issue_title_formatting(self):
        """Test issue title formatting"""
        error = ErrorData(
            id='test-1',
            timestamp=datetime.now().isoformat(),
            level=ErrorLevel.ERROR,
            source=ErrorSource.PYTHON,
            component='TestComponent',
            operation='test_operation',
            message='Test error message',
            metadata={'error_type': 'ValueError'}
        )
        
        title = self.reporter.format_issue_title(error)
        
        self.assertIn('ERROR', title)
        self.assertIn('ValueError', title)
        self.assertIn('TestComponent', title)
        self.assertLessEqual(len(title), 80)
    
    def test_issue_body_formatting(self):
        """Test issue body formatting"""
        error = ErrorData(
            id='test-1',
            timestamp=datetime.now().isoformat(),
            level=ErrorLevel.ERROR,
            source=ErrorSource.PYTHON,
            component='TestComponent',
            operation='test_operation',
            message='Test error message',
            stack_trace='Traceback (most recent call last):\n  File "test.py", line 1',
            details={'key': 'value'},
            metadata={'error_type': 'ValueError'},
            tags=['test', 'component']
        )
        
        body = self.reporter.format_issue_body(error)
        
        # Check for required sections
        self.assertIn('Auto-Generated Error Report', body)
        self.assertIn('test-1', body)
        self.assertIn('Error Message', body)
        self.assertIn('Test error message', body)
        self.assertIn('Stack Trace', body)
        self.assertIn('Additional Details', body)
        self.assertIn('System Information', body)
        self.assertIn('Tags', body)
    
    def test_should_report_error(self):
        """Test error reporting criteria"""
        # Fatal error should be reported
        fatal_error = ErrorData(
            id='test-1',
            timestamp=datetime.now().isoformat(),
            level=ErrorLevel.FATAL,
            source=ErrorSource.PYTHON,
            component='TestComponent',
            operation='test_operation',
            message='Fatal error'
        )
        
        # Error should be reported (since min level is ERROR)
        error = ErrorData(
            id='test-2',
            timestamp=datetime.now().isoformat(),
            level=ErrorLevel.ERROR,
            source=ErrorSource.PYTHON,
            component='TestComponent',
            operation='test_operation',
            message='Error'
        )
        
        # Warning should not be reported (below min level)
        warning = ErrorData(
            id='test-3',
            timestamp=datetime.now().isoformat(),
            level=ErrorLevel.WARNING,
            source=ErrorSource.PYTHON,
            component='TestComponent',
            operation='test_operation',
            message='Warning'
        )
        
        # Note: should_report_error will return False because github_client is not initialized
        # but we can test the level filtering logic by checking the config
        self.assertTrue(ErrorLevel.FATAL >= self.config.min_error_level)
        self.assertTrue(ErrorLevel.ERROR >= self.config.min_error_level)
        self.assertFalse(ErrorLevel.WARNING >= self.config.min_error_level)
    
    def test_create_issue_dry_run(self):
        """Test issue creation in dry run mode"""
        error = ErrorData(
            id='test-1',
            timestamp=datetime.now().isoformat(),
            level=ErrorLevel.ERROR,
            source=ErrorSource.PYTHON,
            component='TestComponent',
            operation='test_operation',
            message='Test error'
        )
        
        # This should not create an actual issue but should return a URL
        # Note: This will return None because github client is not initialized
        # but in dry run mode with a valid client it would return 'dry-run-issue-url'
        result = self.reporter.create_issue(error)
        
        # Check that stats were updated
        self.assertEqual(self.reporter.stats['total_errors_processed'], 1)
    
    def test_get_stats(self):
        """Test statistics retrieval"""
        stats = self.reporter.get_stats()
        
        self.assertIn('total_errors_processed', stats)
        self.assertIn('issues_created', stats)
        self.assertIn('duplicates_skipped', stats)
        self.assertIn('rate_limited', stats)
        self.assertIn('config', stats)
        self.assertIn('current_rate', stats)
        self.assertIn('tracked_fingerprints', stats)


if __name__ == '__main__':
    unittest.main()

File: test/js/test_github_issue_reporter.js
/**
 * Test GitHub Issue Reporter (JavaScript)
 * 
 * Tests the functionality of the GitHub issue reporter in JavaScript.
 */

import { strict as assert } from 'assert';
import {
  GitHubIssueReporter,
  IssueReportConfig,
  getReporter,
  reportError,
  ErrorLevel,
  ErrorSource
} from '../../hallucinate_app/node/github_issue_reporter.js';

describe('GitHubIssueReporter', function() {
  let reporter;
  let config;
  
  beforeEach(function() {
    // Create a test configuration with dry_run enabled
    config = new IssueReportConfig({
      enabled: true,
      githubToken: 'test_token',
      repository: 'test_owner/test_repo',
      minErrorLevel: ErrorLevel.ERROR,
      dryRun: true  // Don't actually create issues during tests
    });
    
    reporter = new GitHubIssueReporter(config);
  });
  
  describe('Configuration', function() {
    it('should create config from environment variables', function() {
      // Set environment variables
      process.env.GITHUB_ISSUE_REPORTER_ENABLED = 'true';
      process.env.GITHUB_TOKEN = 'test_token';
      process.env.GITHUB_REPOSITORY = 'owner/repo';
      process.env.GITHUB_ISSUE_MIN_LEVEL = 'warning';
      process.env.GITHUB_ISSUE_RATE_LIMIT = '5';
      process.env.GITHUB_ISSUE_DRY_RUN = 'true';
      
      const envConfig = new IssueReportConfig();
      
      assert.strictEqual(envConfig.enabled, true);
      assert.strictEqual(envConfig.githubToken, 'test_token');
      assert.strictEqual(envConfig.repository, 'owner/repo');
      assert.strictEqual(envConfig.minErrorLevel, 'warning');
      assert.strictEqual(envConfig.rateLimitPerHour, 5);
      assert.strictEqual(envConfig.dryRun, true);
      
      // Clean up
      delete process.env.GITHUB_ISSUE_REPORTER_ENABLED;
      delete process.env.GITHUB_TOKEN;
      delete process.env.GITHUB_REPOSITORY;
      delete process.env.GITHUB_ISSUE_MIN_LEVEL;
      delete process.env.GITHUB_ISSUE_RATE_LIMIT;
      delete process.env.GITHUB_ISSUE_DRY_RUN;
    });
  });
  
  describe('Error Fingerprinting', function() {
    it('should generate same fingerprint for identical errors', function() {
      const error1 = {
        id: 'test-1',
        timestamp: new Date().toISOString(),
        level: ErrorLevel.ERROR,
        source: ErrorSource.JAVASCRIPT,
        component: 'TestComponent',
        operation: 'test_operation',
        message: 'Test error message'
      };
      
      const error2 = {
        id: 'test-2',
        timestamp: new Date().toISOString(),
        level: ErrorLevel.ERROR,
        source: ErrorSource.JAVASCRIPT,
        component: 'TestComponent',
        operation: 'test_operation',
        message: 'Test error message'
      };
      
      const fingerprint1 = reporter.generateErrorFingerprint(error1);
      const fingerprint2 = reporter.generateErrorFingerprint(error2);
      
      assert.strictEqual(fingerprint1, fingerprint2);
    });
    
    it('should generate different fingerprint for different errors', function() {
      const error1 = {
        id: 'test-1',
        timestamp: new Date().toISOString(),
        level: ErrorLevel.ERROR,
        source: ErrorSource.JAVASCRIPT,
        component: 'TestComponent',
        operation: 'test_operation',
        message: 'Test error message'
      };
      
      const error2 = {
        id: 'test-2',
        timestamp: new Date().toISOString(),
        level: ErrorLevel.WARNING,
        source: ErrorSource.JAVASCRIPT,
        component: 'TestComponent',
        operation: 'test_operation',
        message: 'Different error message'
      };
      
      const fingerprint1 = reporter.generateErrorFingerprint(error1);
      const fingerprint2 = reporter.generateErrorFingerprint(error2);
      
      assert.notStrictEqual(fingerprint1, fingerprint2);
    });
  });
  
  describe('Duplicate Detection', function() {
    it('should detect duplicate errors', function() {
      const fingerprint = 'test_fingerprint';
      
      // Initially not a duplicate
      assert.strictEqual(reporter.isDuplicate(fingerprint), false);
      
      // Mark as reported
      reporter.reportedFingerprints.add(fingerprint);
      reporter.fingerprintTimestamps.set(fingerprint, Date.now());
      
      // Should now be detected as duplicate
      assert.strictEqual(reporter.isDuplicate(fingerprint), true);
    });
  });
  
  describe('Rate Limiting', function() {
    it('should enforce rate limits', function() {
      // Initially should pass rate limit
      assert.strictEqual(reporter.checkRateLimit(), true);
      
      // Add many recent issues
      const now = Date.now();
      for (let i = 0; i < config.rateLimitPerHour; i++) {
        reporter.issueCountPerHour.push(now);
      }
      
      // Should now fail rate limit
      assert.strictEqual(reporter.checkRateLimit(), false);
    });
  });
  
  describe('Issue Title Formatting', function() {
    it('should format issue title correctly', function() {
      const error = {
        id: 'test-1',
        timestamp: new Date().toISOString(),
        level: ErrorLevel.ERROR,
        source: ErrorSource.JAVASCRIPT,
        component: 'TestComponent',
        operation: 'test_operation',
        message: 'Test error message',
        metadata: { errorType: 'ValueError' }
      };
      
      const title = reporter.formatIssueTitle(error);
      
      assert.ok(title.includes('ERROR'));
      assert.ok(title.includes('ValueError'));
      assert.ok(title.includes('TestComponent'));
      assert.ok(title.length <= 80);
    });
  });
  
  describe('Issue Body Formatting', function() {
    it('should format issue body with all sections', function() {
      const error = {
        id: 'test-1',
        timestamp: new Date().toISOString(),
        level: ErrorLevel.ERROR,
        source: ErrorSource.JAVASCRIPT,
        component: 'TestComponent',
        operation: 'test_operation',
        message: 'Test error message',
        stackTrace: 'Error: Test\n    at test.js:1:1',
        details: { key: 'value' },
        metadata: { errorType: 'ValueError' },
        tags: ['test', 'component']
      };
      
      const body = reporter.formatIssueBody(error);
      
      // Check for required sections
      assert.ok(body.includes('Auto-Generated Error Report'));
      assert.ok(body.includes('test-1'));
      assert.ok(body.includes('Error Message'));
      assert.ok(body.includes('Test error message'));
      assert.ok(body.includes('Stack Trace'));
      assert.ok(body.includes('Additional Details'));
      assert.ok(body.includes('System Information'));
      assert.ok(body.includes('Tags'));
    });
  });
  
  describe('Error Reporting Criteria', function() {
    it('should report errors above minimum level', function() {
      const error = {
        id: 'test-1',
        timestamp: new Date().toISOString(),
        level: ErrorLevel.ERROR,
        source: ErrorSource.JAVASCRIPT,
        component: 'TestComponent',
        operation: 'test_operation',
        message: 'Test error'
      };
      
      // Note: shouldReportError will return false because octokit is not initialized
      // but we can verify the config is set correctly
      assert.strictEqual(config.minErrorLevel, ErrorLevel.ERROR);
    });
  });
  
  describe('Create Error from Exception', function() {
    it('should create error object from Error instance', function() {
      const error = new Error('Test error');
      error.stack = 'Error: Test\n    at test.js:1:1';
      
      const errorData = GitHubIssueReporter.createErrorFromException(error, {
        component: 'TestComponent',
        operation: 'test_operation',
        source: ErrorSource.NODEJS,
        level: ErrorLevel.ERROR,
        details: { additionalInfo: 'test' }
      });
      
      assert.strictEqual(errorData.message, 'Test error');
      assert.strictEqual(errorData.component, 'TestComponent');
      assert.strictEqual(errorData.operation, 'test_operation');
      assert.strictEqual(errorData.source, ErrorSource.NODEJS);
      assert.strictEqual(errorData.level, ErrorLevel.ERROR);
      assert.strictEqual(errorData.stackTrace, error.stack);
      assert.ok(errorData.id);
      assert.ok(errorData.timestamp);
    });
  });
  
  describe('Statistics', function() {
    it('should track statistics', function() {
      const stats = reporter.getStats();
      
      assert.ok('totalErrorsProcessed' in stats);
      assert.ok('issuesCreated' in stats);
      assert.ok('duplicatesSkipped' in stats);
      assert.ok('rateLimited' in stats);
      assert.ok('config' in stats);
      assert.ok('currentRate' in stats);
      assert.ok('trackedFingerprints' in stats);
    });
  });
});
